polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tracking/limiting global allocator. Calculates the peak allocation between two checkpoints for
+//! the whole process. Accepts an optional limit and a failure handler which is called if the limit
+//! is overflown.
+
+use core::alloc::{GlobalAlloc, Layout};
+use std::{
+	ptr::null_mut,
+	sync::atomic::{AtomicBool, Ordering},
+};
+
+struct TrackingAllocatorData {
+	lock: AtomicBool,
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn()>>,
+}
+
+impl TrackingAllocatorData {
+	#[inline]
+	fn lock(&self) {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.failure_handler = failure_handler;
+		self.unlock();
+		core::mem::drop(old_handler);
+	}
+
+	fn end_tracking(&mut self) -> isize {
+		self.lock();
+		let peak = self.peak;
+		self.limit = 0;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.unlock();
+		core::mem::drop(old_handler);
+		peak
+	}
+
+	#[inline]
+	fn track(&mut self, alloc: isize) -> bool {
+		self.lock();
+		self.current += alloc;
+		if self.current > self.peak {
+			self.peak = self.current;
+		}
+		let within_limits = self.limit == 0 || self.peak <= self.limit;
+		if within_limits {
+			self.unlock()
+		}
+		within_limits
+	}
+}
+
+static mut ALLOCATOR_DATA: TrackingAllocatorData = TrackingAllocatorData {
+	lock: AtomicBool::new(false),
+	current: 0,
+	peak: 0,
+	limit: 0,
+	failure_handler: None,
+};
```

Comment:
Now that I think about it, we don't have to make this `static mut` now that we have a spinlock in there. It'd be nicer to move the `lock` out of there, make it into `SpinLock<T>` and then this could be just `static ALLOCATOR_DATA: SpinLock<TrackingAllocatorData>`.

Reply:
You mean to put the underlying data into an `UnsafeCell`? Do you mean it's just a more beautiful abstraction (I do agree it is), or do you think it could give some benefits in the sense of performance/safety/whatever?

Reply:
Asking because I believe it would complicate other things quite a bit... E.g., dropping an old failure handler in `end_tracking()`. With the internal lock controlled by `TrackingAllocatorData` itself, it's easy. But when the locking is handled one level higher, it could become a problem.

Reply:
Yeah, tried to implement and stuck into it at once. If we have `ALLOCATOR_DATA: SpinLock<TrackingAllocatorData>` then we call underlying methods like `ALLOCATOR_DATA.lock().end_tracking()`, where `lock()` returns a guard that dereferences to `&mut TrackingAllocatorData` and unlocks when dropped. But if inside `end_tracking()` we need to drop something, it will deallocate and deadlock, as we don't have the possibility to control locking from inside the inner implementation. That works well in ordinary code, but trying to do that in an allocator is tricky :thinking:

Reply:
You could have the method take the guard instead of `&mut self`, and then it could decide itself when to drop the guard, so it could first `take()` whatever it needs to deallocate, drop the guard, and then deallocate it, no?

Reply:
Have you benchmarked to compare the performance with your `Spinlock`, std's `Mutex`, and some existing [spinlock implementation](https://crates.io/crates/spin)? It'd be nice to avoid the extra complexity and `unsafe` usage. Spinlocks also apparently have some [footguns](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html).

Reply:
> In reality, you just trade your own carefully written, checked, and tested unsafe, which is always at hand to fix in case of problems, for some other guy's unsafe, which might be less carefully written, checked, and tested.

I mean, it's generally safer to use code that's been around longer and used by more people (92 million downloads ;), than to write it yourself. But yeah, I've been convinced that here it makes perfect sense to do so.

Reply:
On the other hand, IIRC the `spin` crate is `no_std`, so it shouldn't allocate. I also don't see why a spinlock would call `futex`.  What other guarantees do we need?

Reply:
> On the other hand, IIRC the `spin` crate is `no_std`, so it shouldn't allocate. I also don't see why a spinlock would call `futex`. What other guarantees do we need?

In general personally I don't like pulling in external dependencies for simple things; the whole `spin` crate is 2.5k lines of code, where implementing a spin lock is, like, under ~100 lines of code.

But I realize that not everyone is me and not everyone finds implementing a spin lock trivial (it does contain `unsafe` after all!) so @mrcnski @s0me0ne-unkn0wn I'll leave in up to you if you want to use the `spin` crate for it. (If you are going to use it just make sure to use the `spin::SpinMutex`, because their default implementation is multiplexed over a feature flag, which we most certainly do not want.)

Reply:
It's something that I'd personally use a crate for (also to benefit future readers of our code). It's not much code but easy to get wrong - concepts like memory orderings are not trivial. They probably have comprehensive tests (I hope) which we don't. And to prevent `spin` sneakily adding a syscall in the future or otherwise breaking the guarantees we need, we could pin it to a specific version. Won't block on this though.